### PR TITLE
ci: Migrate from Black and Flake8 to Ruff

### PR DESCRIPTION
**Pull request by barndawgie**
This PR migrates our pre-commit flow from Black and Flake8 to the new Ruff framework, which has already been adopted by Home-Assistant.

The following changes are included:
* Remove Black and Flake8 from pre-commit
* Add ruff configuration to pyproject.toml, using the default configuration with a few key overrides borrowed from https://github.com/home-assistant/core/blob/ca886de3cacdaf59f56411dc91206a4acd0eba46/pyproject.toml#L570
* Migrate codespell config into the consolidated pyproject.toml configuration file
* Include formatting changes made automatically by ruff format (`pre-commit run --all-files`)

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 esphome>=2023.12.0
 pre-commit>=3.4.0
+tomli>=2.0.1
```
Comment (tronikos) [inline requirements.txt:3]: what's this for?

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+[tool.codespell]
+# Check filename spelling in addition to file contents
+check-filenames = ''
+
+[tool.ruff]
+fix = true
+show-fixes = true
+line-length = 120
```
Comment (tronikos) [inline pyproject.toml:8]: remove for now to avoid diffs in files.

**Review by tronikos — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 1 above)
Comment (barndawgie) [reply requirements.txt]: Some doc indicated you needed tomli for ruff to parse your config files but now I can't find that anywhere; removing it.

File: pyproject.toml (hunk 2 above)
Comment (tronikos) [reply pyproject.toml]: Removed

**Review by tronikos — APPROVED** [review]

**Pull request merged**